sycl/test-e2e/Graph/graph_exception_global_device_extension.cpp

```diff
@@ -0,0 +1,146 @@
+// REQUIRES: level_zero, gpu
+// RUN: %{build} -o %t.out
+// RUN: %{run} %t.out
+//
+// The test checks that invalid exception is thrown
+// when trying to use sycl_ext_oneapi_device_global
+// along with Graph.
+
+#include "graph_common.hpp"
+
+using TestProperties = decltype(sycl::ext::oneapi::experimental::properties{});
+
+sycl::ext::oneapi::experimental::device_global<int, TestProperties>
+    MemcpyDeviceGlobal;
+sycl::ext::oneapi::experimental::device_global<int, TestProperties>
+    CopyDeviceGlobal;
+
+enum OperationPath { Explicit, RecordReplay, Shortcut };
+
+template <OperationPath PathKind> void test() {
+  queue Q;
+  int MemcpyWrite = 42, CopyWrite = 24, MemcpyRead = 1, CopyRead = 2;
+
+  exp_ext::command_graph Graph{Q.get_context(), Q.get_device()};
+
+  Graph.begin_recording(Q);
+
+  // Copy from device globals before having written anything.
+  std::error_code ExceptionCode = make_error_code(sycl::errc::success);
+  try {
+    if constexpr (PathKind == OperationPath::Shortcut) {
+      Q.memcpy(&MemcpyRead, MemcpyDeviceGlobal);
+    }
+    if constexpr (PathKind == OperationPath::RecordReplay) {
+      Q.submit([&](handler &CGH) {
+        return CGH.memcpy(&MemcpyRead, MemcpyDeviceGlobal);
+      });
+    }
+    if constexpr (PathKind == OperationPath::Explicit) {
+      Graph.add([&](handler &CGH) {
+        return CGH.memcpy(&MemcpyRead, MemcpyDeviceGlobal);
+      });
```

Comment:
This could be throwing an error because of using the explicit API while a Graph is in recording mode (i.e https://github.com/reble/llvm/pull/283), rather than anything to do with an incompatible extension. Thinking we might need to guard the `Graph.begin_recording(Q);` & `  Graph.end_recording(Q);` calls with `if constexpr (PathKind != OperationPath::Explicit)`

Think the same point affects the unittests as well

Reply:
I corrected this in test-e2e and unitests.